src/shared_ptr.hpp

```diff
@@ -1,8 +1,109 @@
-# pragma once
+#pragma once
+
+#include <functional>
 
 template<typename T>
 class shared_ptr
 {
 public:
-  explicit shared_ptr(T*){}
+  shared_ptr()
+    : mValue(nullptr)
+    , mCtrlBlock(nullptr)
+  {
+  }
+
+  explicit shared_ptr(T* aPtr)
+    : mCtrlBlock(new control_block(1, 0))
+  {
+    mValue = aPtr;
+  }
+
+  explicit shared_ptr(T* aPtr, std::function<void(T*)> aDeleter)
+    : mCtrlBlock(new control_block(1, 0, aDeleter))
+  {
+    mValue = aPtr;
+  }
+
+  shared_ptr(const shared_ptr& aPtr) noexcept
+  {
+    if (aPtr.mCtrlBlock)
+    {
+      aPtr.mCtrlBlock->StrongRef++;
+    }
+
+    mValue = aPtr.mValue;
+    mCtrlBlock = aPtr.mCtrlBlock;
+  }
+
+  shared_ptr& operator=(const shared_ptr<T>& aPtr) noexcept
+  {
+    if (mCtrlBlock == aPtr.mCtrlBlock && mValue == aPtr.mValue)
+    {
+      return *this;
+    }
+
+    // assuming if mValue != nullptr means that mCtrlBlock can't be nullptr as well
+    if (mValue)
+    {
+      if (mCtrlBlock->StrongRef <= 1)
+      {
+        mCtrlBlock->Deleter ? mCtrlBlock->Deleter(mValue) : delete mValue;
```

Comment:
if (mCtrlBlock->StrongRef <= 1)
      {
        mCtrlBlock->Deleter ? mCtrlBlock->Deleter(mValue) : delete mValue;
        ...
this code is duplicated at least in two places, so it should be a private method to avoid code duplication